pulser-core/pulser/sequence/_seq_drawer.py

```diff
@@ -670,6 +534,306 @@ def phase_str(phi: float) -> str:
+        sampling_rate: Sampling rate of the effective pulse used by
+            the solver. If present, plots the effective pulse alongside the
+            input pulse.
+        draw_phase_area: Whether phase and area values need to be shown
+            as text on the plot, defaults to False. If `draw_phase_curve=True`,
+            phase values are ommited.
+        draw_interp_pts: When the sequence has pulses with waveforms of
+            type InterpolatedWaveform, draws the points of interpolation on
+            top of the respective waveforms (defaults to True).
+        draw_phase_shifts: Whether phase shift and reference information
+            should be added to the plot, defaults to False.
+        draw_register: Whether to draw the register before the pulse
+            sequence, with a visual indication (square halo) around the qubits
+            masked by the SLM, defaults to False.
+        draw_input: Draws the programmed pulses on the channels, defaults
+            to True.
+        draw_modulation: Draws the expected channel output, defaults to
+            False. If the channel does not have a defined 'mod_bandwidth', this
+            is skipped unless 'draw_input=False'.
+        draw_phase_curve: Draws the changes in phase in its own curve (ignored
+            if the phase doesn't change throughout the channel).
+    """
+
+    def phase_str(phi: float) -> str:
+        """Formats a phase value for printing."""
+        value = (((phi + np.pi) % (2 * np.pi)) - np.pi) / np.pi
+        if value == -1:
+            return r"$\pi$"
+        elif value == 0:
+            return "0"  # pragma: no cover - just for safety
+        else:
+            return rf"{value:.2g}$\pi$"
+
+    n_channels = len(seq.declared_channels)
+    if not n_channels:
+        raise RuntimeError("Can't draw an empty sequence.")
+
+    # Sample the sequence and get the data to plot
+    sampled_seq = sample(seq)
+    data = gather_data(sampled_seq)
+
+    # Gather additional data for sequence specific drawing
+    for ch, sch in seq._schedule.items():
+        interp_pts: defaultdict[str, list[list[float]]] = defaultdict(list)
+
+        for slot in sch:
+            if slot.ti == -1 or slot.type in ["target", "delay"]:
+                continue
+
+            pulse = cast(Pulse, slot.type)
+            for wf_type in ["amplitude", "detuning"]:
+                wf = getattr(pulse, wf_type)
+                if isinstance(wf, InterpolatedWaveform):
+                    pts = wf.data_points
+                    pts[:, 0] += slot.ti
+                    interp_pts[wf_type] += pts.tolist()
+
+        if interp_pts:
+            data[ch].interp_pts = dict(interp_pts)
+
+    if hasattr(seq, "_measurement"):
+        data["measurement"] = seq._measurement
+
+    # Boxes for qubit and phase text
+    area_ph_box = dict(boxstyle="round", facecolor="ghostwhite", alpha=0.7)
+    ph_box = dict(boxstyle="round", facecolor="ghostwhite")
+
```

Comment:
As all of this can be moved after / put in `draw_channel_content`, I suggest we do so and have `gather_data` called in `draw_channel_content`. That way `channel_draw_content` only takes a `SequenceSamples` and not a `SequenceSamples` and a `data` dict.

Reply:
Sorry I have not been very clear, I meant that since we have (if you move the `measurement` part in `_draw_channel_content`)
```
    data = gather_data(sampled_seq)
    for ch, sch in seq._schedule.items():
        interp_pts: defaultdict[str, list[list[float]]] = defaultdict(list)

        for slot in sch:
            if slot.ti == -1 or slot.type in ["target", "delay"]:
                continue

            pulse = cast(Pulse, slot.type)
            for wf_type in ["amplitude", "detuning"]:
                wf = getattr(pulse, wf_type)
                if isinstance(wf, InterpolatedWaveform):
                    pts = wf.data_points
                    pts[:, 0] += slot.ti
                    interp_pts[wf_type] += pts.tolist()

        if interp_pts:
            data[ch].interp_pts = dict(interp_pts)
    # Boxes for qubit and phase text
    area_ph_box = dict(boxstyle="round", facecolor="ghostwhite", alpha=0.7)
    ph_box = dict(boxstyle="round", facecolor="ghostwhite")
    (fig_reg, fig, ch_axes, data) = draw_channel_content(
        data,
        sampled_seq,
        seq.register if draw_register else None,
        sampling_rate,
        draw_input,
        draw_modulation,
        draw_phase_curve,
    )
```
And that `interp_pts` of data is unused in `draw_channel_content`, you can re-write this as
```
    data = gather_data(sampled_seq)
    (fig_reg, fig, ch_axes, data) = draw_channel_content(
        data,
        sampled_seq,
        seq.register if draw_register else None,
        sampling_rate,
        draw_input,
        draw_modulation,
        draw_phase_curve,
    )
    for ch, sch in seq._schedule.items():
        interp_pts: defaultdict[str, list[list[float]]] = defaultdict(list)

        for slot in sch:
            if slot.ti == -1 or slot.type in ["target", "delay"]:
                continue

            pulse = cast(Pulse, slot.type)
            for wf_type in ["amplitude", "detuning"]:
                wf = getattr(pulse, wf_type)
                if isinstance(wf, InterpolatedWaveform):
                    pts = wf.data_points
                    pts[:, 0] += slot.ti
                    interp_pts[wf_type] += pts.tolist()

        if interp_pts:
            data[ch].interp_pts = dict(interp_pts)
    # Boxes for qubit and phase text
    area_ph_box = dict(boxstyle="round", facecolor="ghostwhite", alpha=0.7)
    ph_box = dict(boxstyle="round", facecolor="ghostwhite")
```
And as you have the structure
```
    data = gather_data(sampled_seq)
    (fig_reg, fig, ch_axes, data) = draw_channel_content(
        data,
        sampled_seq,
        ...
    )
...
```
for `draw_sequence` and `draw_samples`, I propose you modify `draw_channel_content` such that it calls `gather_data` (first operation it does after testing validity of `sampled_seq`) and that it only takes `sampled_seq` (no longer need `data` in the list of attributes).
Such that in the end you only have
```

    (fig_reg, fig, ch_axes, data) = draw_channel_content(
        sampled_seq,
        ...
    )
...
```

Reply:
oh, got it!